Enabling JIT support breaks compilation (in many ways) on Linux

Affects 32-Bit (chroot) and 64-Bit, tested on Debian Testing/Stretch.

Dyngen appears to be experiencing some serious issues, possibly due to assumptions about compiler behaviour and changes in newer compiler versions.

GCC Version: `(Debian 6.3.0-5) 6.3.0 20170124`

Failure : "Unsupported Relocation ... (4)" (Dyngen)
========================================

x86_64: `dyngen: unsupported AMD64 relocation (4)`

i386: `dyngen: unsupported i386 relocation (4)`

Disabling PIE generation "fixed this.

Failure 2: Frame Pointer Required
===========================

```
In file included from ../kpx_cpu/src/cpu/jit/basic-dyngen-ops.cpp:21:0:
./sysdeps.h: In function ‘uint32 do_opt_bswap_32(uint32)’:
./sysdeps.h:188:22: error: frame pointer required, but reserved
 static inline uint32 do_opt_bswap_32(uint32 x)
```

"Assuaged" by passing `-fomit-frame-pointer` to GCC

Failure 3: Memory corruption in Dyngen
================================

```
./dyngen -o basic-dyngen-ops.hpp obj/basic-dyngen-ops.o
*** Error in `./dyngen': malloc(): memory corruption (fast): 0x0000000079f72b60 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x70bcb)[0x7f0174b10bcb]
/lib/x86_64-linux-gnu/libc.so.6(+0x76fa6)[0x7f0174b16fa6]
/lib/x86_64-linux-gnu/libc.so.6(+0x793c1)[0x7f0174b193c1]
/lib/x86_64-linux-gnu/libc.so.6(__libc_malloc+0x54)[0x7f0174b1ad94]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x91ad4)[0x7f01753eaad4]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x9ae3d)[0x7f01753f3e3d]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(__cxa_demangle+0x7a)[0x7f01753f40fa]
./dyngen[0x7804bd2c]
./dyngen[0x7804b9c6]
./dyngen[0x7804bce3]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7f0174ac02b1]
./dyngen[0x78048e6a]
======= Memory map: ========
78000000-7814e000 rwxp 00000000 00:11 162322680                          /usr/local/src/macemu/SheepShaver/src/Unix/dyngen
79f5a000-79f8c000 rw-p 00000000 00:00 0                                  [heap]
7f0170000000-7f0170021000 rw-p 00000000 00:00 0 
7f0170021000-7f0174000000 ---p 00000000 00:00 0 
7f0174aa0000-7f0174c35000 r-xp 00000000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
7f0174c35000-7f0174e34000 ---p 00195000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
7f0174e34000-7f0174e38000 r--p 00194000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
7f0174e38000-7f0174e3a000 rw-p 00198000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
7f0174e3a000-7f0174e3e000 rw-p 00000000 00:00 0 
7f0174e3e000-7f0174e54000 r-xp 00000000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
7f0174e54000-7f0175053000 ---p 00016000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
7f0175053000-7f0175054000 r--p 00015000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
7f0175054000-7f0175055000 rw-p 00016000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
7f0175055000-7f0175158000 r-xp 00000000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
7f0175158000-7f0175357000 ---p 00103000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
7f0175357000-7f0175358000 r--p 00102000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
7f0175358000-7f0175359000 rw-p 00103000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
7f0175359000-7f01754cb000 r-xp 00000000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
7f01754cb000-7f01756cb000 ---p 00172000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
7f01756cb000-7f01756d5000 r--p 00172000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
7f01756d5000-7f01756d7000 rw-p 0017c000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
7f01756d7000-7f01756db000 rw-p 00000000 00:00 0 
7f01756db000-7f01756fe000 r-xp 00000000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
7f017589f000-7f01758a3000 rw-p 00000000 00:00 0 
7f01758f9000-7f01758fd000 rw-p 00000000 00:00 0 
7f01758fd000-7f01758fe000 r--p 00022000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
7f01758fe000-7f01758ff000 rw-p 00023000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
7f01758ff000-7f0175900000 rw-p 00000000 00:00 0 
7ffec07d6000-7ffec07f8000 rw-p 00000000 00:00 0                          [stack]
7ffec07fa000-7ffec07fc000 r--p 00000000 00:00 0                          [vvar]
7ffec07fc000-7ffec07fe000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
Makefile:225: recipe for target 'basic-dyngen-ops.hpp' failed
make: *** [basic-dyngen-ops.hpp] Aborted
make: *** Deleting file 'basic-dyngen-ops.hpp'
```

If the problem is just with dyngen, you could try making the build use the
pre-generated cpp and hpp files that are generated by that tool. They've
been checked into the source repository under
SheepShaver/src/Unix/dyngen_precompiled
for x86 and x86_64.

Right now, only the OS X Xcode 8 build is set up to use them. Ideally an
option to autogen.sh could be added to use those pre-built files instead of
building & running the dyngen tool to auto-generate them. Patches welcome!

Having said that, if we could make dyngen work with modern GCC - that would
be great as well! It sounds like you at least got some ways towards making
it work. If you figure out all the problems, please send a pull request
with your changes!

On Sat, Feb 4, 2017 at 2:43 PM, Roman Hargrave <notifications@github.com>
wrote:

> Affects 32-Bit (chroot) and 64-Bit, tested on Debian Testing/Stretch.
>
> Dyngen appears to be experiencing some serious issues, possibly due to
> assumptions about compiler behaviour and changes in newer compiler versions.
>
> GCC Version: (Debian 6.3.0-5) 6.3.0 20170124
> Failure : "Unsupported Relocation ... (4)" (Dyngen)
>
> x86_64: dyngen: unsupported AMD64 relocation (4)
>
> i386: dyngen: unsupported i386 relocation (4)
>
> Disabling PIE generation "fixed this.
> Failure 2: Frame Pointer Required
>
> In file included from ../kpx_cpu/src/cpu/jit/basic-dyngen-ops.cpp:21:0:
> ./sysdeps.h: In function ‘uint32 do_opt_bswap_32(uint32)’:
> ./sysdeps.h:188:22: error: frame pointer required, but reserved
>  static inline uint32 do_opt_bswap_32(uint32 x)
>
> "Assuaged" by passing -fomit-frame-pointer to GCC
> Failure 3: Memory corruption in Dyngen
>
> ./dyngen -o basic-dyngen-ops.hpp obj/basic-dyngen-ops.o
> *** Error in `./dyngen': malloc(): memory corruption (fast): 0x0000000079f72b60 ***
> ======= Backtrace: =========
> /lib/x86_64-linux-gnu/libc.so.6(+0x70bcb)[0x7f0174b10bcb]
> /lib/x86_64-linux-gnu/libc.so.6(+0x76fa6)[0x7f0174b16fa6]
> /lib/x86_64-linux-gnu/libc.so.6(+0x793c1)[0x7f0174b193c1]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_malloc+0x54)[0x7f0174b1ad94]
> /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x91ad4)[0x7f01753eaad4]
> /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x9ae3d)[0x7f01753f3e3d]
> /usr/lib/x86_64-linux-gnu/libstdc++.so.6(__cxa_demangle+0x7a)[0x7f01753f40fa]
> ./dyngen[0x7804bd2c]
> ./dyngen[0x7804b9c6]
> ./dyngen[0x7804bce3]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7f0174ac02b1]
> ./dyngen[0x78048e6a]
> ======= Memory map: ========
> 78000000-7814e000 rwxp 00000000 00:11 162322680                          /usr/local/src/macemu/SheepShaver/src/Unix/dyngen
> 79f5a000-79f8c000 rw-p 00000000 00:00 0                                  [heap]
> 7f0170000000-7f0170021000 rw-p 00000000 00:00 0
> 7f0170021000-7f0174000000 ---p 00000000 00:00 0
> 7f0174aa0000-7f0174c35000 r-xp 00000000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
> 7f0174c35000-7f0174e34000 ---p 00195000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
> 7f0174e34000-7f0174e38000 r--p 00194000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
> 7f0174e38000-7f0174e3a000 rw-p 00198000 00:11 160056693                  /lib/x86_64-linux-gnu/libc-2.24.so
> 7f0174e3a000-7f0174e3e000 rw-p 00000000 00:00 0
> 7f0174e3e000-7f0174e54000 r-xp 00000000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
> 7f0174e54000-7f0175053000 ---p 00016000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
> 7f0175053000-7f0175054000 r--p 00015000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
> 7f0175054000-7f0175055000 rw-p 00016000 00:11 162068732                  /lib/x86_64-linux-gnu/libgcc_s.so.1
> 7f0175055000-7f0175158000 r-xp 00000000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
> 7f0175158000-7f0175357000 ---p 00103000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
> 7f0175357000-7f0175358000 r--p 00102000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
> 7f0175358000-7f0175359000 rw-p 00103000 00:11 160056697                  /lib/x86_64-linux-gnu/libm-2.24.so
> 7f0175359000-7f01754cb000 r-xp 00000000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
> 7f01754cb000-7f01756cb000 ---p 00172000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
> 7f01756cb000-7f01756d5000 r--p 00172000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
> 7f01756d5000-7f01756d7000 rw-p 0017c000 00:11 162066075                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22
> 7f01756d7000-7f01756db000 rw-p 00000000 00:00 0
> 7f01756db000-7f01756fe000 r-xp 00000000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
> 7f017589f000-7f01758a3000 rw-p 00000000 00:00 0
> 7f01758f9000-7f01758fd000 rw-p 00000000 00:00 0
> 7f01758fd000-7f01758fe000 r--p 00022000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
> 7f01758fe000-7f01758ff000 rw-p 00023000 00:11 160056689                  /lib/x86_64-linux-gnu/ld-2.24.so
> 7f01758ff000-7f0175900000 rw-p 00000000 00:00 0
> 7ffec07d6000-7ffec07f8000 rw-p 00000000 00:00 0                          [stack]
> 7ffec07fa000-7ffec07fc000 r--p 00000000 00:00 0                          [vvar]
> 7ffec07fc000-7ffec07fe000 r-xp 00000000 00:00 0                          [vdso]
> ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
> Makefile:225: recipe for target 'basic-dyngen-ops.hpp' failed
> make: *** [basic-dyngen-ops.hpp] Aborted
> make: *** Deleting file 'basic-dyngen-ops.hpp'
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cebix/macemu/issues/107>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABE8HSHiBFOxRqH31cFajWgHjb3Bc7zks5rZNTggaJpZM4L3Qop>
> .
>

--- a/SheepShaver/src/Unix/Makefile.in
+++ b/SheepShaver/src/Unix/Makefile.in
@@ -26,6 +26,7 @@ CPUSRCS = @CPUSRCS@
 MONSRCS = @MONSRCS@
 PERL = @PERL@
 USE_DYNGEN = @USE_DYNGEN@
+USE_DYNGEN_PRECOMPILED = @USE_DYNGEN_PRECOMPILED@
 DYNGENSRCS = @DYNGENSRCS@
 DYNGEN_CC = @DYNGEN_CC@
 DYNGEN_CFLAGS = @DYNGEN_CFLAGS@
@@ -209,9 +210,28 @@ DYNGEN    = dyngen$(EXEEXT)
 ifeq ($(USE_DYNGEN),yes)
 DYNGENDEPS = basic-dyngen-ops.hpp ppc-dyngen-ops.hpp
 
+ifeq ($(USE_DYNGEN_PRECOMPILED),yes)
+basic-dyngen-ops.hpp: dyngen_precompiled/basic-dyngen-ops.hpp basic-dyngen-ops-x86_32.hpp basic-dyngen-ops-x86_64.hpp
+	cp -f $< $@
+basic-dyngen-ops-x86_32.hpp: dyngen_precompiled/basic-dyngen-ops-x86_32.hpp
+	cp -f $< $@
+basic-dyngen-ops-x86_64.hpp: dyngen_precompiled/basic-dyngen-ops-x86_64.hpp
+	cp -f $< $@
+ppc-dyngen-ops.hpp: dyngen_precompiled/ppc-dyngen-ops.hpp ppc-dyngen-ops-x86_32.hpp ppc-dyngen-ops-x86_64.hpp
+	cp -f $< $@
+ppc-dyngen-ops-x86_32.hpp: dyngen_precompiled/ppc-dyngen-ops-x86_32.hpp
+	cp -f $< $@
+ppc-dyngen-ops-x86_64.hpp: dyngen_precompiled/ppc-dyngen-ops-x86_64.hpp
+	cp -f $< $@
+else
 # Only GCC is supported for generating synthetic opcodes
 $(DYNGEN): $(DYNGENOBJS)
 	$(DYNGEN_CC) -o $@ $(LDFLAGS) $(DYNGENOBJS)
+basic-dyngen-ops.hpp: $(OBJ_DIR)/basic-dyngen-ops.o $(DYNGEN)
+	./$(DYNGEN) -o $@ $<
+ppc-dyngen-ops.hpp: $(OBJ_DIR)/ppc-dyngen-ops.o $(DYNGEN)
+	./$(DYNGEN) -o $@ $<
+endif
 
 $(OBJ_DIR)/%.dgo : %.c
 	$(DYNGEN_CC) -xc $(CPPFLAGS) $(DEFS) $(DYNGEN_CFLAGS) -c $< -o $@
@@ -221,14 +241,10 @@ $(OBJ_DIR)/%.dgo : %.cpp
 $(OBJ_DIR)/basic-dyngen.o: basic-dyngen-ops.hpp
 $(OBJ_DIR)/basic-dyngen-ops.o: $(kpxsrcdir)/cpu/jit/basic-dyngen-ops.cpp
 	$(DYNGEN_CC) $(CPPFLAGS) $(DEFS) $(DYNGEN_CXXFLAGS) $(DYNGEN_OP_FLAGS) -c $< -o $@
-basic-dyngen-ops.hpp: $(OBJ_DIR)/basic-dyngen-ops.o $(DYNGEN)
-	./$(DYNGEN) -o $@ $<
 
 $(OBJ_DIR)/ppc-dyngen.o: ppc-dyngen-ops.hpp
 $(OBJ_DIR)/ppc-dyngen-ops.o: $(kpxsrcdir)/cpu/ppc/ppc-dyngen-ops.cpp basic-dyngen-ops.hpp
 	$(DYNGEN_CC) $(CPPFLAGS) $(DEFS) $(DYNGEN_CXXFLAGS) $(DYNGEN_OP_FLAGS) -c $< -o $@
-ppc-dyngen-ops.hpp: $(OBJ_DIR)/ppc-dyngen-ops.o $(DYNGEN)
-	./$(DYNGEN) -o $@ $<
 
 $(OBJ_DIR)/sheepshaver_glue.o $(OBJ_DIR)/ppc-cpu.o $(OBJ_DIR)/ppc-decode.o $(OBJ_DIR)/ppc-translate.o $(OBJ_DIR)/ppc-jit.o: basic-dyngen-ops.hpp ppc-dyngen-ops.hpp
 endif

--- a/SheepShaver/src/Unix/configure.ac
+++ b/SheepShaver/src/Unix/configure.ac
@@ -46,7 +46,7 @@ AC_ARG_WITH(gtk,            [  --with-gtk              use GTK user interface [d
    esac],
   [WANT_GTK="gtk2 gtk"])
 AC_ARG_WITH(mon,            [  --with-mon              use mon as debugger [default=yes]], [WANT_MON=$withval], [WANT_MON=yes])
-AC_ARG_WITH(dgcc,           [  --with-dgcc=COMPILER    use C++ COMPILER to compile synthetic opcodes], [DYNGEN_CC=$withval])
+AC_ARG_WITH(dgcc,           [  --with-dgcc=COMPILER    use C++ COMPILER to compile synthetic opcodes or 'precompiled'], [DYNGEN_CC=$withval])
 
 AC_ARG_WITH(bincue,   
   AS_HELP_STRING([--with-bincue], [Allow cdrom image files in bin/cue mode]))
@@ -1432,6 +1432,17 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#if ! (__GNUC__ >= 3)
                   [AC_MSG_RESULT(yes); HAVE_GCC30=yes],
                   [AC_MSG_RESULT(no)])
 
+dnl Check for GCC 4.0 or higher.
+HAVE_GCC40=no
+AC_MSG_CHECKING(for GCC 4.0 or higher)
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#if ! (__GNUC__ >= 4)
+                                     # error gcc < 4
+                                     typedef syntax error;
+                                     #endif
+                                   ]])],
+                  [AC_MSG_RESULT(yes); HAVE_GCC40=yes],
+                  [AC_MSG_RESULT(no)])
+
 dnl Check for ICC.
 AC_MSG_CHECKING(for ICC)
 HAVE_ICC=no
@@ -1504,6 +1515,7 @@ if [[ "x$EMULATED_PPC" = "xyes" ]]; then
 
   dnl Enable JIT compiler, if possible
   if [[ "x$WANT_JIT" = "xyes" ]]; then
+    ac_cv_use_dyngen_precompiled=no
     AC_CACHE_CHECK([whether dyngen can be used],
       ac_cv_use_dyngen, [
       case $host_cpu:$ac_cv_object_format in
@@ -1534,20 +1546,30 @@ if [[ "x$EMULATED_PPC" = "xyes" ]]; then
       esac
       dnl Check for a suitable synthetic opcodes compiler (icc is faking itself as gcc 3.2.2)
       if [[ -z "$DYNGEN_CC" ]]; then
-        if [[ "x$HAVE_GCC27" = "xyes" -a "x$HAVE_ICC" = "xno" ]]; then
+        if [[ "x$HAVE_GCC27" = "xyes" -a "x$HAVE_GCC40" = "xno" -a "x$HAVE_ICC" = "xno" ]]; then
           DYNGEN_CC=$CXX
-        else
-          for p in /usr/bin /usr/local/bin /usr/freeware/bin; do
-            gxx="$p/g++"
-            if [[ -x "$gxx" ]]; then
-              DYNGEN_CC="$gxx"
-            fi
-          done
+        elif command -v g++ >/dev/null; then
+          vers=`g++ -dumpversion`
+          function version { echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }'; }
+          if [[ $(version $vers) -ge $(version "2.7.0") ]] && [[ $(version $vers) -lt $(version "4.0.0") ]]; then
+            DYNGEN_CC="$gxx"
+          fi
         fi
       fi
-      if [[ -z "$DYNGEN_CC" ]]; then
-        ac_cv_use_dyngen=no
+      if [[ -z "$DYNGEN_CC" -o "x$DYNGEN_CC" = "xprecompiled" ]]; then
+        case $host_cpu in
+          i?86)
+            ac_cv_use_dyngen_precompiled=yes
+            ;;
+          x86_64)
+            ac_cv_use_dyngen_precompiled=yes
+            ;;
+          *)
+            ac_cv_use_dyngen=no
+            ;;
+        esac
       fi
+      if [[ "x$ac_cv_use_dyngen_precompiled" = "xyes" ]]; then DYNGEN_CC=precompiled; fi
     ])
     if [[ "x$ac_cv_use_dyngen" = "xyes" ]]; then
       case $host_cpu in
@@ -1564,10 +1586,17 @@ if [[ "x$EMULATED_PPC" = "xyes" ]]; then
         ;;
       esac
       have_dyngen_gcc3=no
-      case "x`$DYNGEN_CC -dumpversion`" in
-      x[12].*) ;;
-      x*) have_dyngen_gcc3=yes ;;
-      esac
+      if [[ "$DYNGEN_CC" != "precompiled" ]]; then
+        case "x`$DYNGEN_CC -dumpversion`" in
+        x[12].*) ;;
+        x*) have_dyngen_gcc3=yes ;;
+        esac
+      else
+        case "x`$CXX -dumpversion`" in
+        x[12].*) ;;
+        x*) have_dyngen_gcc3=yes ;;
+        esac
+      fi
       if [[ "x$have_dyngen_gcc3" = "xyes" ]]; then
         DYNGEN_OP_FLAGS="$DYNGEN_OP_FLAGS -fno-align-functions"
       else
@@ -1651,6 +1680,7 @@ AC_TRANSLATE_DEFINE(HAVE_LINKER_SCRIPT, "$ac_cv_linker_script_works",
 dnl Generate Makefile.
 AC_SUBST(PERL)
 AC_SUBST(USE_DYNGEN, [$ac_cv_use_dyngen])
+AC_SUBST(USE_DYNGEN_PRECOMPILED, [$ac_cv_use_dyngen_precompiled])
 AC_SUBST(DYNGENSRCS)
 AC_SUBST(DYNGEN_CC)
 AC_SUBST(DYNGEN_CFLAGS)

